core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -1120,6 +1135,52 @@ object ZManaged {
       ZManaged.environment.flatMap(f)
   }
 
+  /**
+   * A `FinalizerRef` describes the finalizers associated with a scope.
+   *
+   */
+  trait FinalizerRef[R] {
+
+    /**
+     * Adds a finalizer to the finalizers associated with this scope. If the
+     * finalizers associated with this scope have already been run this
+     * finalizer will be run immediately.
+     */
+    def add(finalizer: Exit[Any, Any] => URIO[R, Any]): URIO[R, Unit]
+
+    /**
+     * Runs the finalizers associated with this scope. After this any
+     * finalizers added to this scope will be run immediately.
+     */
+    def run(exit: Exit[Any, Any]): URIO[R, Unit]
+  }
+
+  object FinalizerRef {
+
+    /**
+     * Constructs a new `FinalizerRef` with the specified initial finalizer.
+     */
+    def make[R](initial: Exit[Any, Any] => URIO[R, Any]): UIO[FinalizerRef[R]] =
+      Ref.make[Either[Exit[Any, Any], Exit[Any, Any] => URIO[R, Any]]](Right(initial)).map { ref =>
```

Comment:
Would it be possible for the `Right` case here to be a `Set[Exit[Any, Any] => URIO[R, Any]]`? We can then have, in addition to `run`, `runPar` and combinators for manipulating the finalizers (`remove`, `replace`, etc). At that point, I don't think we'd need `finalizerRefInternal` anymore.

Needless to say, this is really nice!

Reply:
Thanks!

I think we could definitely make this a set. I'm a little worried about safety of some of the combinators though. `runPar` should be fine though it seems like at least right now it leads to finalizers not running in certain situations. I think `remove` or `replace` may not be safe because even if you remove or replace a finalizer it may not have gotten attached yet so you may not actually be able to remove the finalizer. Perhaps we could maintain a separate set of "deactivated" finalizers and change `add` to check the deactivated finalizers.

Reply:
I added a combinator to run a specified finalizer, safely removing it from the `FinalizerRef` and used this to fix the issue with `scope`. Still thinking about how to safely add more combinators to `FinalizerRef` to implement the remaining methods that use `finalizerRefInternal` (`ZManaged.switchable`, `ZStream.catchAllCause`, and `Stream.flatMap`) but may make sense to merge this if we think it is correct and follow up on that.

Reply:
Yes, makes sense. Will review again today and merge.